monolith/src/monolith.rs

```diff
@@ -0,0 +1,226 @@
+
+        let tmp = y ^ !y_rot_1 & y_rot_2 & y_rot_3;
+        (tmp >> 7) | (tmp << 1)
+    }
+
+    pub fn final_s_box(y: u8) -> u8 {
+        debug_assert_eq!(y >> 7, 0); // must be a 7-bit value
+
+        let y_rot_1 = (y >> 6) | (y << 1);
+        let y_rot_2 = (y >> 5) | (y << 2);
+
+        let tmp = (y ^ !y_rot_1 & y_rot_2) & 0x7F;
+        ((tmp >> 6) | (tmp << 1)) & 0x7F
+    }
+
+    fn instantiate_lookup1() -> Vec<u16> {
+        (0..=u16::MAX)
+            .map(|i| {
+                let lo = (i >> 8) as u8;
+                let hi = i as u8;
+                ((Self::s_box(lo) as u16) << 8) | Self::s_box(hi) as u16
+            })
+            .collect()
+    }
+
+    fn instantiate_lookup2() -> Vec<u16> {
+        (0..(1 << 15))
+            .map(|i| {
+                let lo = (i >> 8) as u8;
+                let hi: u8 = i as u8;
+                ((Self::final_s_box(lo) as u16) << 8) | Self::s_box(hi) as u16
+            })
+            .collect()
+    }
+
+    fn random_field_element(shake: &mut Shake128Reader) -> Mersenne31 {
+        let mut val = get_random_u32(shake);
+        while val >= Mersenne31::ORDER_U32 {
+            val = get_random_u32(shake);
+        }
+
+        Mersenne31::from_canonical_u32(val)
+    }
+
+    fn init_shake() -> Shake128Reader {
+        let mut shake = Shake128::default();
+        shake.update("Monolith".as_bytes());
+        shake.update(&[WIDTH as u8, NUM_ROUNDS as u8]);
+        shake.update(&Mersenne31::ORDER_U32.to_le_bytes());
+        shake.update(&[8, 8, 8, 7]);
+        shake.finalize_xof()
+    }
+
+    fn instantiate_round_constants() -> Vec<[Mersenne31; WIDTH]> {
+        let mut shake = Self::init_shake();
+
+        vec![[Mersenne31::ZERO; WIDTH]; NUM_ROUNDS - 1]
+            .iter()
+            .map(|arr| arr.map(|_| Self::random_field_element(&mut shake)))
+            .collect()
+    }
+
+    pub fn concrete(
+        &self,
+        state: &mut [Mersenne31; WIDTH],
+        round_constants: Option<&[Mersenne31; WIDTH]>,
+    ) {
+        *state = self.mds.permute(*state);
+
+        if let Some(round_constants) = round_constants {
+            for (x, rc) in state.iter_mut().zip(round_constants.iter()) {
+                *x += *rc;
+            }
+        }
+    }
+
+    pub fn bricks(state: &mut [Mersenne31; WIDTH]) {
+        // Feistel Type-3
+        for (x, x_mut) in (state.to_owned()).iter().zip(state.iter_mut().skip(1)) {
+            let x_mut_u64 = &mut x_mut.as_canonical_u64();
+            let x_u64 = x.as_canonical_u64();
+            // Every time at bricks the input is technically a u32, so we tell the compiler
+            let mut tmp_square = (x_u64 & 0xFFFFFFFF_u64) * (x_u64 & 0xFFFFFFFF_u64);
+            tmp_square %= Mersenne31::ORDER_U64;
+            *x_mut_u64 = (*x_mut_u64 & 0xFFFFFFFF_u64) + (tmp_square & 0xFFFFFFFF_u64);
+            *x_mut_u64 %= Mersenne31::ORDER_U64;
+            *x_mut = Mersenne31::from_canonical_u64(*x_mut_u64);
```

Comment:
Could we simplify it to this?
```suggestion
            *x_mut += x.square();
```
Maybe it's suboptimal since there may be a bit of overlapping work between the two reductions (squaring and addition), but should be cheaper than the two mod instructions.

Reply:
Oh wow haha, yeah that should be fine!